### PR TITLE
AWS IAM support

### LINKED ISSUE #40: Add AWS IAM support
# This is a (Bug Report / Feature Proposal)
Feature Proposal
## Description
AWS IAM is very much essential and used almost before accessing any other service.

For bug reports:
* What went wrong?
* What did you expect should have happened?
* What was the config you used?
* What stacktrace or error message from your provider did you see?

For feature proposals:
* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.
AWS IAM allows user to create groups, attach policies to them, create users, etc
* If there is additional config how would it look

Similar or dependent issues:
* 

## Additional Data

* ***Operating System***:
* ***Stack Trace***:
* ***Provider Error messages***:

### PR DESCRIPTION
**Pull request by raju249**
## What did you implement:

Closes #40 

## How did you implement it:
Added AWS IAM basic functions using aws node sdk.

## How can we verify it:
Run the AWS IAM example.

## Todos:

- [x] Write tests
- [x] Write documentation
- [ ] Fix linting errors
- [ ] Make sure code coverage hasn't dropped
- [ ] Provide verification config / commands / resources
- [ ] Enable "Allow edits from maintainers" for this PR
- [ ] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: examples/security/aws-iam.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+const nodeCloud = require('../../lib/');
+var assert = require('assert')
```
Comment (rajikaimal) [inline examples/security/aws-iam.js]: can you change `var` to `const`

**Review by rajikaimal — CHANGES_REQUESTED** [review]

**Comment by raju249** [comment]
@rajikaimal 
Acknowledged changes
Please review when you have time. :smile:

File: lib/security/aws-iam.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	* @param {object} options - {apiVersion}
+	*/
+
+	constructor(aws, options) {
+		this._AWS = aws;
+		this._apiVersion = options.apiVersion;
+		this._iam = new this._AWS.IAM({ apiVersion: this._apiVersion });
+	}
+	/**
+   * Create IAM Group
```
Comment (rajikaimal) [inline lib/security/aws-iam.js]: Can you reformat comments, some indentations are not inline with others.

**Review by rajikaimal — CHANGES_REQUESTED** [review]

File: lib/security/aws-iam.js (hunk 2 above)
Comment (raju249) [reply lib/security/aws-iam.js]: Sure,
Aligned the remaining comments with proper alignment.
Thanks @rajikaimal

**Comment by raju249** [comment]
Ah wait,
Some issue with my editor, they are not changed, let me add a commit with requested changes

**Comment by raju249** [comment]
@rajikaimal Did the needful.
Thanks for the changes 😄

File: examples/security/aws-iam.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+		console.log('Attaching Group Policy');
+
+		iam
+			.attachGroupPolicy(GroupPolicyParams)
+			.then((res) => {
+				console.log(res);
+
+				console.log('Detaching Group Policy');
+
+				iam
```
Comment (rajikaimal) [inline examples/security/aws-iam.js]: Let's move this block into `then` block of first promise. Don't use promises within promises.

File: examples/security/aws-iam.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+					.detachGroupPolicy(GroupPolicyParams)
+					.then((res) => {
+						console.log(res);
+
+						console.log("Deleting Group");
+
+						iam
```
Comment (rajikaimal) [inline examples/security/aws-iam.js]: Same with this promise block too.

**Review by rajikaimal — CHANGES_REQUESTED** [review]

File: examples/security/aws-iam.js (hunk 3 above)
Comment (raju249) [reply examples/security/aws-iam.js]: Actually, before attaching policy to a group, the group should be created, nodejs runs code asynchronously, due to which I need to make sure the previous step in completed.
Removing nested promises ensures this ?

**Comment by raju249** [comment]
@rajikaimal  
Thanks a lot, now the code for example looks cleaner :smile:

File: examples/security/aws-iam.js [hunk 5]
```diff
@@ -16,41 +16,16 @@ const GroupPolicyParams = {
...
+        return iam.attachGroupPolicy(GroupPolicyParams);
+    })
+    .then((res) => {
+        return iam.detachGroupPolicy(GroupPolicyParams);
+    })
+    .then((res) => {
+        return iam.deleteGroup(GroupParams);
+    })
+    .then((res) => {
+        console.log(res)
```
Comment (rajikaimal) [inline examples/security/aws-iam.js]: Add missing semi colon here :), this looks solid!

**Review by rajikaimal — CHANGES_REQUESTED** [review]

**Comment by rajikaimal** [comment]
@raju249 no problem 😄

File: examples/security/aws-iam.js [hunk 6]
```diff
@@ -28,4 +28,7 @@ iam.createGroup(GroupParams)
     })
     .then((res) => {
         console.log(res)
-    });
+    })
+		.catch((err) => {
+			console.log(err);
```
Comment (rajikaimal) [inline examples/security/aws-iam.js]: make this `console.error`.

**Comment by raju249** [comment]
Did both the changes, anything else to include @rajikaimal 
Sorry that you had to suggest lot of changes :sweat_smile:

**Comment by rajikaimal** [comment]
LGTM! @raju249 no problem, great work.

**Review by rajikaimal — APPROVED** [review]

**Comment by rajikaimal** [comment]
Can you add this to documentation as well. You can refer docs folder.

**Comment by raju249** [comment]
Yeah sure @rajikaimal

**Pull request merged**